### ISSUE TITLE
Most-recently used (MRU) list of notebooks

### ISSUE THREAD

**Issue by roblevy** [jupyter/notebook#2653]
Every launch of Jupyter currently starts (for me) with a tedious browse through my folder system, looking for the `.ipynb` file I was working on yesterday.

99 times out of 100 I want to pick up where I left off, choosing from a list of, say, the last 10 notebooks I opened.

This might be an issue with my own workflow (In Windows > Start > Jupyter Notebook) but I don't *think* it is. Apologies if I'm wrong.

**Comment by takluyver** [comment]
Thanks - I think this is a sensible feature request, but it may be a while before anyone gets to it.

**Comment by anshumusaddi** [comment]
Hello!
We are a group of undergrads from India who are looking to solve issues in open source repositories as a course project. Some of us(including me), being avid users of Jupyter Notebook suggested that we choose an issue in this repository.

We are highly motivated and doing work on PR#2968. Meanwhile, we would start working on this as we are a team of 10 people. Can we start implementing it with your guidance?

**Comment by gnestor** [comment]
@anshumusaddi Certainly! Mention me if you run into any issues or would like me to review your work 👍

**Comment by prof-lupin** [comment]
Hey @gnestor ! Thanks for the reply. I'm with @anshumusaddi .
We have made quite some progress with the planning and, w.r.t. implementation, we are thinking of employing a min heap to store the list of recent files at the client side. The MRU tab will display the files by accessing that only. We have demarcated the task of handling the backend and front end among us. We would love to hear your opinion and suggestions on this. Hoping for some insights from you side.

**Comment by takluyver** [comment]
I suspect there's no need to worry about using a heap for this - it probably only needs something like 10-20 notebooks, so performance isn't going to be a concern.

I might lean towards storing the list on the server side, so if you connect to the same server from different machines, your MRU list is shared. But if it seems more feasible to store it client side, that's still useful in many cases, and it could always be extended later.

Have you considered what counts as using a notebook? Opening/running cells/saving?

**Comment by prof-lupin** [comment]
Hey @takluyver ! You made some valid points. We will now be implementing a list instead of a heap and are considering opening a notebook as using it. We are inclined towards storing the list on the client side as it will give the user their personalised list rather than a complete list or else the data in server will be very large and highly unscalable. Also, there is no way to differentiate a user from another to give them access to their MRU list from amidst the MRU lists of all the users.

**Comment by takluyver** [comment]
The `notebook` package (the repository this issue is on) is meant to be single user - you can give multiple people access, but it's not really designed for that. [JupyterHub](https://jupyterhub.readthedocs.io/en/stable/) is the multi-user layer on top of the notebook, and that does distinguish between different users and can handle giving each their own space on the server.

**Comment by prof-lupin** [comment]
Duly noted, @takluyver . We have shifted our design and plan to work on the server side. Also, I am attaching the UI we will be implementing. We have created a 'Recent' tab which will display the most recent notebooks sorted according to the time. Please tell us your opinions on this and we will be happy to oblige.
![IMG-20191016-WA0044](https://user-images.githubusercontent.com/33233393/67096107-083c2c00-f1d5-11e9-87bc-717b7e551199.jpg)

**Comment by takluyver** [comment]
That design looks good to me.

**Comment by prof-lupin** [comment]
Hello, @gnestor @takluyver ! Hope you are doing well. One of my teammates has submitted a PR  #(PR #5051 ) consisting our final work on this project. We passed the integration testing on Travis CI not long ago and hope that you all approve of this as well. Although, if there still are some significant nooks then we implore you to tell us so that we may list them out in the next phase of this project. Thank you!

**Comment by prof-lupin** [comment]
Hey @takluyver @gnestor ! It was a great experience contributing to the Jupyter community. We would really appreciate it if you could take a look at our PR and reiterate your reviews. Part of the journey is the end and we would love to know how we fared at that. Thank you.
